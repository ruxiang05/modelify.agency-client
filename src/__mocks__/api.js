const jobMock = {
  status: 'in progress',
  _id: '5c97b9cd0328e6b2aa76a361',
  model: '5c686eb24cb1b33456b0be59',
  agent: '5c686e904cb1b33456b0be58',
  title: 'Test',
  date: '2019-05-26T00:00:00.000Z',
  address: 'address',
  pay: 100,
  description: 'some description',
};

const api = {
  jobs: {
    getJobs: jest.fn(() => Promise.resolve({
      jobs: [jobMock],
    })),
    acceptJob: jest.fn(() => Promise.resolve({
      job: jobMock,
    })),
    completeJob: jest.fn(() => Promise.resolve({
      job: {
        ...jobMock,
        status: 'complete',
      },
    })),
    declineJob: jest.fn(() => Promise.resolve({
      job: {
        ...jobMock,
        status: 'declined',
      },
    })),
    addJob: jest.fn(() => Promise.resolve()),
  },
  googleCalendar: {
    addEvent: jest.fn(() => Promise.resolve()),
    authorize: jest.fn(() => Promise.resolve({ message: 'success', token: 'test' })),
    getAuthURL: jest.fn(() => Promise.resolve({ redirect: '/redirect' })),
  },
  agents: {
    getModels: jest.fn(() => Promise.resolve({
      agent: {
        agentInfo: {
          models: [
            {
              _id: '5ca24d217736960017cae87c',
              agentInfo: { models: [] },
              email: 'jane@doe.com',
              modelInfo: {
                address: '',
                characteristics: {
                  eyes: '',
                  hair: '',
                  skin: '',
                  dateOfBirth: '',
                },
                measurements: {
                  chest: '',
                  height: '',
                  hips: '',
                  waist: '',
                  weight: '',
                },
                name: 'Jane Doe',
                phoneNumber: '07423360809',
                role: 'model',
              },
            },
          ],
        },
      },
    })),
    addModel: jest.fn(() => Promise.resolve()),
  },
  chats: {
    getMessages: jest.fn(() => Promise.resolve({
      messages: [
        {
          message: 'hi',
          user: '1',
        },
        {
          message: 'hello',
          user: '2',
        },
      ],
    })),
    getChats: jest.fn(() => Promise.resolve({
      chats: [
        {
          id: '2t3457345',
          job: {},
          messages: [],
          users: ['1', '2'],
        },
      ],
    })),
  },
  users: {
    signup: jest.fn(() => Promise.resolve({ ok: true })),
    login: jest.fn(() => Promise.resolve({ token: 'abc' })),
    editProfile: jest.fn(() => Promise.resolve({ token: 'abc' })),
  },
};

module.exports = api;
