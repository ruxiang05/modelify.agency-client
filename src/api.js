const url = process.env.REACT_APP_API_URL;

const api = {
  users: {
    signup: data => fetch(`${url}/users/signup`, {
      method: 'POST',
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json',
      },
    }),
    login: data => fetch(`${url}/users/login`, {
      method: 'POST',
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json',
      },
    }),
    editProfile: (token, data) => fetch(`${url}/users/edit`, {
      method: 'PUT',
      body: JSON.stringify(data),
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    })
      .then((res, err) => {
        if (err) {
          return err;
        }
        return res.json();
      })
      .catch(err => err),
  },
  agents: {
    getModels: token => fetch(`${url}/agents/`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    })
      .then((res, err) => {
        if (err) {
          return err;
        }
        return res.json();
      })
      .catch(err => err),
    addModel: (token, data) => fetch(`${url}/agents/add-model`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'content-type': 'application/json',
      },
      body: JSON.stringify(data),
    })
      .then((res, err) => {
        if (err) {
          return err;
        }
        return res.json();
      })
      .catch(err => err),
  },
  jobs: {
    addJob: (token, data) => fetch(`${url}/jobs/create`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'content-type': 'application/json',
      },
      body: JSON.stringify(data),
    })
      .then((res, err) => {
        if (err) {
          return err;
        }
        return res.json();
      })
      .catch(err => err),
    getJobs: token => fetch(`${url}/jobs/all`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    })
      .then((res, err) => {
        if (err) {
          return err;
        }
        return res.json();
      })
      .catch(err => err),
  },
};

export default api;
