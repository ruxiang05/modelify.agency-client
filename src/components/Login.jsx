import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import FormInput from './FormInput';
import { UserContext } from '../contexts/userContext';
import api from '../api';
import { setToken } from '../auth';
import '../styles/login.scss';
import Logo from './Logo';

class Login extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const { name, value } = event.target;
    this.setState(prevState => ({ ...prevState, [name]: value }));
  }

  handleSubmit(event) {
    const user = this.state;
    const { history } = this.props;
    const { updateUser } = this.context;
    event.preventDefault();
    api.users
      .login(user)
      .then((data) => {
        setToken(data.token);
        updateUser();
        history.push('/');
      })
      .catch(error => error);
  }

  render() {
    const { email, password } = this.state;
    return (
      <div className="page centered">
        <Logo />
        <form onSubmit={this.handleSubmit}>
          <FormInput
            type="email"
            name="email"
            title="Email"
            value={email}
            placeholder="Enter your email"
            handleChange={this.handleChange}
          />
          <FormInput
            type="password"
            name="password"
            title="Password"
            value={password}
            placeholder="Enter your password"
            handleChange={this.handleChange}
          />
          <div className="form-submit">
            <input type="submit" value="Log in" />
          </div>
        </form>
        <div className="to-sign-up">
          <p>OR</p>
          <Link to="/signup">Sign up</Link>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  history: PropTypes.shape({}).isRequired,
};

Login.contextType = UserContext;
export default Login;
