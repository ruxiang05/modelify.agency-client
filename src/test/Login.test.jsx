/* Uses methods and/or components from react, prop-types, history and enzyme */
import React from 'react';
import { shallow } from 'enzyme';
import PropTypes from 'prop-types';
import { createBrowserHistory } from 'history';
import Login from '../components/Login';
import api from '../api';

jest.mock('../api.js');

let wrapper;

const contextTypes = {
  updateUser: PropTypes.func,
};

const context = {
  updateUser: jest.fn(),
};

const history = createBrowserHistory({
  basename: '/',
  keyLength: 0,
});

const formSubmit = {
  email: 'email@email.com',
  password: '123',
};

beforeAll(() => {
  Login.contextTypes = contextTypes;
});

beforeEach(() => {
  wrapper = shallow(<Login history={history} />, { context });
});

describe('Login', () => {
  it('should render Login', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle input change', async () => {
    const event = {
      target: {
        name: 'name',
        value: 'name',
      },
    };
    wrapper.instance().handleChange(event);
    await wrapper.update();
    expect(wrapper.state().name).toBe('name');
  });

  it('should handle form submit', async () => {
    const event = {
      preventDefault: jest.fn(),
    };
    wrapper.setState({ ...formSubmit });
    await wrapper.update();
    wrapper.instance().handleSubmit(event);
    expect(api.users.login).toHaveBeenCalled();
  });

  it('should error if signup fails', async () => {
    const event = {
      preventDefault: jest.fn(),
    };
    api.users.login = jest
      .fn()
      .mockImplementationOnce(() => Promise.reject(new Error('api error')));
    wrapper.setState({ ...formSubmit });
    await wrapper.update();
    wrapper.instance().handleSubmit(event);
    expect(api.users.login).toHaveBeenCalled();
  });
});
