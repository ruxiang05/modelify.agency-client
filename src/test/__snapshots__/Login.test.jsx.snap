// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render Login 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Logo />,
        <form
          onSubmit={[Function]}
        >
          <FormInput
            handleChange={[Function]}
            name="email"
            placeholder="Enter your email"
            title="Email"
            type="email"
            value=""
          />
          <FormInput
            handleChange={[Function]}
            name="password"
            placeholder="Enter your password"
            title="Password"
            type="password"
            value=""
          />
          <div
            className="form-submit"
          >
            <input
              type="submit"
              value="Log in"
            />
          </div>
        </form>,
        <div
          className="to-sign-up"
        >
          <p>
            OR
          </p>
          <Link
            replace={false}
            to="/signup"
          >
            Sign up
          </Link>
        </div>,
      ],
      "className": "page centered",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FormInput
              handleChange={[Function]}
              name="email"
              placeholder="Enter your email"
              title="Email"
              type="email"
              value=""
            />,
            <FormInput
              handleChange={[Function]}
              name="password"
              placeholder="Enter your password"
              title="Password"
              type="password"
              value=""
            />,
            <div
              className="form-submit"
            >
              <input
                type="submit"
                value="Log in"
              />
            </div>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleChange": [Function],
              "name": "email",
              "placeholder": "Enter your email",
              "title": "Email",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleChange": [Function],
              "name": "password",
              "placeholder": "Enter your password",
              "title": "Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                type="submit"
                value="Log in"
              />,
              "className": "form-submit",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "type": "submit",
                "value": "Log in",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              OR
            </p>,
            <Link
              replace={false}
              to="/signup"
            >
              Sign up
            </Link>,
          ],
          "className": "to-sign-up",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "OR",
            },
            "ref": null,
            "rendered": "OR",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Sign up",
              "replace": false,
              "to": "/signup",
            },
            "ref": null,
            "rendered": "Sign up",
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Logo />,
          <form
            onSubmit={[Function]}
          >
            <FormInput
              handleChange={[Function]}
              name="email"
              placeholder="Enter your email"
              title="Email"
              type="email"
              value=""
            />
            <FormInput
              handleChange={[Function]}
              name="password"
              placeholder="Enter your password"
              title="Password"
              type="password"
              value=""
            />
            <div
              className="form-submit"
            >
              <input
                type="submit"
                value="Log in"
              />
            </div>
          </form>,
          <div
            className="to-sign-up"
          >
            <p>
              OR
            </p>
            <Link
              replace={false}
              to="/signup"
            >
              Sign up
            </Link>
          </div>,
        ],
        "className": "page centered",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FormInput
                handleChange={[Function]}
                name="email"
                placeholder="Enter your email"
                title="Email"
                type="email"
                value=""
              />,
              <FormInput
                handleChange={[Function]}
                name="password"
                placeholder="Enter your password"
                title="Password"
                type="password"
                value=""
              />,
              <div
                className="form-submit"
              >
                <input
                  type="submit"
                  value="Log in"
                />
              </div>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleChange": [Function],
                "name": "email",
                "placeholder": "Enter your email",
                "title": "Email",
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleChange": [Function],
                "name": "password",
                "placeholder": "Enter your password",
                "title": "Password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  type="submit"
                  value="Log in"
                />,
                "className": "form-submit",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "type": "submit",
                  "value": "Log in",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                OR
              </p>,
              <Link
                replace={false}
                to="/signup"
              >
                Sign up
              </Link>,
            ],
            "className": "to-sign-up",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "OR",
              },
              "ref": null,
              "rendered": "OR",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Sign up",
                "replace": false,
                "to": "/signup",
              },
              "ref": null,
              "rendered": "Sign up",
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "updateUser": [MockFunction],
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
